coda-src/auth2/clog.1
.TH "CLOG" "1" "25 April 2005" "Coda Distributed File System" ""

.SH NAME
clog \- Authenticate to the Coda distributed file system
.SH SYNOPSIS

\fBclog\fR [ \fB-pipe\fR ] [ \fB-test\fR ] [ \fB-host \fIhostname\fB\fR ] [ \fB-tofile \fIfile\fB\fR ] [ \fB-fromfile \fIfile\fB\fR ] \fB[username][@realm]\fR

.SH "DESCRIPTION"
.PP
This command enables users to get tokens to use for authenticated
communication with a Coda File System.  The given name and a password
are passed to an \fBAuthentication
Server\fR\&. The \fBAuthentication Server\fR returns
a set of tokens encoding the chosen identity if the password is correct.
\fBclog\fR passes these tokens to the
\fBvenus\fR process, which acts as the users agent for file
manipulation. These tokens have a limited lifetime, becoming invalid after
25 hours. If an attempt is made to operate on a Coda File System object
while the user is not authenticated, the user will assume the privileges
of the \fBAnonymous\fR user.
.PP
\fBclog\fR accepts the following arguments:
.TP
\fB-pipe\fR
For use in shell scripts, the password is read from stdin.
This option is also enabled when stdin is not a tty.
.TP
\fB-test\fR
Test the clog-venus token passing code.
.TP
\fB-host\fR
Specify the host where the authentication server is
running.
.TP
\fB-tofile\fR
After obtaining a token from the authentication server,
write it to a file.
.TP
\fB-fromfile\fR
Instead of connecting to the authentication server, read a
previously saved token from a file and pass it to venus.
.TP
\fB[username][@realm]\fR
Coda user name you wish to be identified as within the
specified realm. If the username is not specified, clog tries to
use the local login name. If the Coda realm name is not specified
clog falls back to using the default realm name as specified in
the venus.conf configuration file.
.SH "BUGS"
.PP
.TP 0.2i
\(bu
The on-disk tokens are not encrypted.
.TP 0.2i
\(bu
The \fB-pipe\fR option should be dropped, there
already is an isatty(stdin) test which is sufficient for most
cases.
.TP 0.2i
\(bu
The \fB-test\fR option should also test the
reading and writing code of token files.
.SH "SEE ALSO"
.PP
\fBcunlog\fR(1),
\fBvenus\fR(8)
.SH "AUTHORS"
.PP
.TP 0.2i
\(bu
1987, Adapted from AFS-2
.TP 0.2i
\(bu
Maria R. Ebling, 1990, Created man page


NEWS
Incompatible RVM update (Coda client)
-------------------------------------

2008/02/06 RVM version 39
    Move store identifier from per-volume to global rvg struct.

2005/06/07 RVM version 38
    Added realm::realmid (as we can't use &realm_object on 64-bit systems)

2005/06/16 RVM version 37
    Removed fsdb::device and fsobj::inode



coda-src/auth2/clog.c
/* BLURB gpl

                           Coda File System
                              Release 6

          Copyright (c) 1987-2008 Carnegie Mellon University
                  Additional copyrights listed below

This  code  is  distributed "AS IS" without warranty of any kind under
the terms of the GNU General Public Licence Version 2, as shown in the
file  LICENSE.  The  technical and financial  contributors to Coda are
listed in the file CREDITS.

                        Additional copyrights

#*/

/*
                         IBM COPYRIGHT NOTICE

                          Copyright (C) 1986
             International Business Machines Corporation
                         All Rights Reserved

This  file  contains  some  code identical to or derived from the 1986
version of the Andrew File System ("AFS"), which is owned by  the  IBM
Corporation.   This  code is provided "AS IS" and IBM does not warrant
that it is free of infringement of  any  intellectual  rights  of  any
third  party.    IBM  disclaims  liability of any kind for any damages
whatsoever resulting directly or indirectly from use of this  software
or  of  any  derivative work.  Carnegie Mellon University has obtained
permission to  modify,  distribute and sublicense this code,  which is
based on Version 2  of  AFS  and  does  not  contain  the features and
enhancements that are part of  Version 3 of  AFS.  Version 3 of AFS is
commercially   available   and  supported  by   Transarc  Corporation,
Pittsburgh, PA.

*/

/*

	log -- tell Venus your password

*/

#ifdef __cplusplus
extern "C" {
#endif

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <sys/param.h>
#include <sys/types.h>
#include <stdio.h>
#include <pwd.h>
#include <unistd.h>
#include <stdlib.h>
#include "coda_string.h"

#include <lwp/lwp.h>
#include <rpc2/rpc2.h>
#include <auth2.h>

#include "avenus.h"
#include "auser.h"
#include "tokenfile.h"
#include "codaconf.h"

#ifdef __cplusplus
}
#endif

#include <parse_realms.h>

#ifndef O_BINARY
#define O_BINARY 0
#endif

void printusage(void)
{
		    fprintf(stderr,
                            "Usage clog [-q] [-test] [-host authserver]"
                            "\t[-kerberos4 | -kerberos5 | -coda]\n"
                            "\t[-tofile <file>] [-fromfile <file>]\n"
			    "\t[-as username] [Coda username][@realm]\n");
}

int main(int argc, char **argv)
{
    EncryptedSecretToken    sToken;
    RPC2_Integer            authmethod = AUTH_METHOD_CODAUSERNAME;
    int                     verbose = 1;
    int                     interactive = 1;
    ClearToken		    cToken;
    EncryptedSecretToken    testSTok;
    ClearToken		    testCTok;
    char *hostname=NULL;
    char *username=NULL;
    const char *realm = NULL;
    long rc;
    int i;
    int testing = 0;
    char *tofile   = NULL;
    char *fromfile = NULL;
    char *runas = NULL;

/* Make intelligent default decisions, depending on how we were built..
		-- Troy <hozer@drgw.net> */
#ifdef HAVE_KRB4
    authmethod = AUTH_METHOD_KERBEROS4;
#endif
#ifdef HAVE_KRB5
    authmethod = AUTH_METHOD_KERBEROS5;
#endif 

    i = 1;
    while (i < argc) {
	    if (U_GetAuthMethod(argv[i], &authmethod)) {
		    i++;
		    continue;
	    } else if ( strcmp(argv[i], "-?") == 0 ||
                        strcmp(argv[i], "-h") == 0 ||
                        strcmp(argv[i], "--help") == 0 ) {
		    printusage();
		    exit(0);
	    } else if ( strcmp(argv[i], "-test") == 0 ) {
		    testing =1;
		    i++;
	    } else if ( strcmp(argv[i], "-nokinit") == 0 ) {
		    interactive = 0;
		    i++;
	    }  else if ( strcmp(argv[i], "-q") == 0 ) {
		    verbose = 0;
		    i++;
	    }  else if ( strcmp(argv[i], "-pipe") == 0 ) {
		    /* obsolete option -pipe used to mainly reduce verbosity */
		    interactive = 0;
		    verbose = 0;
		    i++;
	    }  else if ( strcmp(argv[i], "-tofile") == 0 ) {
		    i++;
		    if (i >= argc) {
			    fprintf(stderr, "Missing file to write token to\n");
			    printusage();
			    exit(1);
		    }
                    tofile = argv[i++];
	    }  else if ( strcmp(argv[i], "-fromfile") == 0 ) {
		    i++;
		    if (i >= argc) {
			    fprintf(stderr,"Missing file to read token from\n");
			    printusage();
			    exit(1);
		    }
                    fromfile = argv[i++];
	    } else if ( strncmp(argv[i], "-h", 2) == 0) {
		    i++;
		    if (i >= argc) {
			    fprintf(stderr, "Missing host\n");
			    printusage();
			    exit(1);
		    }
		    hostname = argv[i];
		    i++;
	    }  else if ( strcmp(argv[i], "-as") == 0 ) {
		    i++;
		    if (i >= argc) {
			    fprintf(stderr, "Missing -as username\n");
			    printusage();
			    exit(1);
		    }
		    runas = argv[i];
		    i++;
	    } else if ( i == argc-1 ) {
                    username = argv[i];
		    SplitRealmFromName(username, &realm);
                    i++;
	    }
	    /* still coming: 
	       -e ".............": newpag, authenticate, exec
	       -u unsafe, tell Venus to use uid instead of pag.
	    */
	    else {
		    fprintf(stderr, "Wrong argument: %s\n", argv[i]);
		    printusage();
		    exit(1);
	    }

    }

#ifndef __CYGWIN32__
    if (runas) {
	struct passwd *pw = getpwnam(runas);
	if (pw) 
	    setuid(pw->pw_uid);
    }
#endif
    
    if (!username || *username == '\0') {
#ifdef __CYGWIN32__
	username = getlogin();	 
#else
	struct passwd *pw = getpwuid(geteuid());
	if (pw) {
	    username=pw->pw_name;
	}
#endif
    }

    codaconf_init("venus.conf");
    codaconf_init("auth2.conf");

    CODACONF_STR(realm, "realm", NULL);

    if (!username || !realm) {
	    fprintf (stderr, "Can't figure out your username or realm.\n");
	    fprintf (stderr, "Try \"clog user[@realm]\"\n");
	    exit (1);
    }

    if (!isatty(0)) {
	verbose = 0;
	interactive = 0;
    }

    if (verbose)
	printf("username: %s@%s\n", username, realm);

    U_InitRPC();

    if (fromfile) {
        ReadTokenFromFile(fromfile, &cToken, sToken);
    } else {
	struct RPC2_addrinfo *srvs = U_GetAuthServers(realm, hostname);
	rc = U_Authenticate(srvs, authmethod, username, strlen(username)+1,
			    &cToken, sToken, verbose, interactive);
	RPC2_freeaddrinfo(srvs);
	if (rc != 0) {
	    fprintf (stderr, "Invalid login (%s).\n", RPC2_ErrorMsg(rc));
	    exit (1);
	}
    }

    if (testing)
	    printf ("Sending token to venus\n");

    if (tofile)
	WriteTokenToFile(tofile, &cToken, sToken);

    if (U_SetLocalTokens(0, &cToken, sToken, realm))
	printf("Local login only, could not contact venus\n");

    if (testing) {
	    printf("Getting tokens back from venus\n");
	    if (U_GetLocalTokens(&testCTok, testSTok, realm) < 0)
		    perror("U_GetLocalTokens");

	    printf("Comparing clear token\n");
	    if(memcmp(&cToken, &testCTok, sizeof(ClearToken)) != 0) {
		    printf("Bad ClearToken\n");
	    }
	    printf("Comparing secret token\n");
	    if(memcmp(sToken, testSTok, sizeof(EncryptedSecretToken)) != 0) {
		printf("Bad SecretToken\n");
	    }

	    fprintf(stderr,"net order:\n");		
	    fprintf(stderr,"\tAuthHandle = %d\n",testCTok.AuthHandle);
	    fprintf(stderr,"\tViceId = %d\n",testCTok.ViceId);
	    fprintf(stderr,"\tBeginTimestamp = %d\n",testCTok.BeginTimestamp);
	    fprintf(stderr,"\tEndTimestamp = %d\n",testCTok.EndTimestamp);
	    U_NetToHostClearToken(&testCTok);
	    fprintf(stderr,"host order:\n");
	    fprintf(stderr,"\tAuthHandle = %d\n",testCTok.AuthHandle);
	    fprintf(stderr,"\tViceId = %d\n",testCTok.ViceId);
	    fprintf(stderr,"\tBeginTimestamp = %d\n",testCTok.BeginTimestamp);
	    fprintf(stderr,"\tEndTimestamp = %d\n",testCTok.EndTimestamp);

	    printf("Done ! ! ! !\n");
    }
    return 0;
}

